src/mono/wasm/runtime/dotnet-crypto-worker.js

```diff
@@ -3,9 +3,14 @@
 
 var ChannelWorker = {
     _impl: class {
+        // LOCK states
+        get LOCK_UNLOCKED() { return 0; } // 0 means the lock is unlocked
+        get LOCK_OWNED() { return 2; } // 2 means the ChannelWorker owns the lock
+
         // BEGIN ChannelOwner contract - shared constants.
         get STATE_IDX() { return 0; }
         get MSG_SIZE_IDX() { return 1; }
+        get LOCK_IDX() { return 2; }
```

Comment:
This is fine, and probably simplifies things, but I'm surprised the state machine isn't handling it, since it uses atomics for transitions.  Do you know how the state machine is getting desynced?

Reply:
I haven't worked out the exact way the race condition happens. What I do know is that when the race happens, the `ChannelWorker` reads `0` for the `MSG_SIZE`. The only way that can happen is that the `ChannelWorker` runs twice in succession, and it is reading the `0` it wrote itself to the `MSG_SIZE` field in the previous page:

https://github.com/dotnet/runtime/blob/54f104a16489b7a2fa5b9661e8030aa23aa494b0/src/mono/wasm/runtime/dotnet-crypto-worker.js#L69-L70

I've verified this by changing that line to `-1`, and when the race happens, the `ChannelWorker` reads `-1` for the `MSG_SIZE`.

Then, when the race condition happens, I logged what the total amount of data received by the `ChannelWorker` was. And when the test fails, the total amount is always `1024` chars (one page size) less than when the test passes. Upping the test from one million `a`s to ten million `a`s, and running the test in a loop on a 2-core machine seems to consistently repro the problem after 5-10 iterations of the loop. (I set the loop for 50, and it never reached the full 50 before reproing.)

My guessing here is that it isn't a simple race condition, but a combination of race conditions that need to happen in a specific order for the code to get into this state.

The "write" side:

https://github.com/dotnet/runtime/blob/54f104a16489b7a2fa5b9661e8030aa23aa494b0/src/mono/wasm/runtime/crypto-worker.ts#L122-L152

The "read" side:
https://github.com/dotnet/runtime/blob/54f104a16489b7a2fa5b9661e8030aa23aa494b0/src/mono/wasm/runtime/dotnet-crypto-worker.js#L51-L73

The first race is at the bottom of the methods - the "write" side doesn't wait to be notified, but instead spins until it reads a new "state". That means the write side can get unblocked before the "read" side calls ".wait". 

Then another race is between lines 139-140 of the "write" and 71-72 of the "read". The "write" side can change the state in between the "read" side moving to 'state = AWAIT' and calling ".wait". So now the "read" side doesn't wait on line 72 (because the state is no longer 'AWAIT'). It then loops around and reads another page of data, sets the state, and then calls ".wait". Meanwhile, the "write" side is after 139, but before 140. Now it runs to call "notify" to wake up the "read" side one more time - and the "read" side now reads `0` for `MSG_SIZE`.

The part I haven't worked out is how that causes the skipping of a page. But given that you can see these two sides running concurrently, and causing unintended state transitions, adding the lock to ensure only one side is in the "critical section" at a time seems like the most appropriate way to fix it.